### PR TITLE
resolve: Some macro resolution refactoring

### LINKED ISSUE #53269: Panic when trying to `use` macro
The compiler panicked

I tried this code:  
[compiler-crash tag on rust-nix](https://gitlab.com/jD91mZM2/rust-nix/tree/compiler-crash).
`cargo test`

I expected to see this happen: A normal compilation error  
Instead, this happened: The rust compiler panicked

## Meta

`rustc --version --verbose`:
```
rustc 1.30.0-nightly (7e8ca9f8b 2018-08-03)
binary: rustc
commit-hash: 7e8ca9f8bd8325398e76bc30ac09aab138bbb127
commit-date: 2018-08-03
host: x86_64-unknown-linux-gnu
release: 1.30.0-nightly
LLVM version: 7.0
```
2018 edition of rust

Backtrace: https://gist.github.com/0bf4ae25ef77940633fda40257b16cab

### LINKED ISSUE #53512: ICE when `use`ing macro defined in standard library
On nightly `1.30.0-nightly (33b923fd4 2018-08-18)`:

If you try to compile

```rust
use std::assert;

fn main() {
    assert!(true);
}
```

you get

```text
error[E0432]: unresolved import `std::assert`
 --> src/main.rs:1:5
  |
1 | use std::assert;
  |     ^^^^^^^^^^^ no `assert` in the root

thread 'main' panicked at 'librustc_resolve/macros.rs:909: inconsistent resolution for a macro', librustc/util/bug.rs:47:26
note: Run with `RUST_BACKTRACE=1` for a backtrace.
error: aborting due to previous error

For more information about this error, try `rustc --explain E0432`.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.30.0-nightly (33b923fd4 2018-08-18) running on x86_64-unknown-linux-gnu

note: compiler flags: -C codegen-units=1 -C debuginfo=2 --crate-type bin

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `playground`.

To learn more, run the command again with --verbose.
```

On playground: https://play.rust-lang.org/?gist=5f24c022a2d8fd3f2256a88a6c92e8a1&version=nightly&mode=debug&edition=2015

### PR DESCRIPTION
**Pull request by petrochenkov**
Work towards completing https://github.com/rust-lang/rust/pull/50911#issuecomment-411605393

The last commit also fixes https://github.com/rust-lang/rust/issues/53269 by not using `def_id()` on `Def::Err` and also fixes https://github.com/rust-lang/rust/issues/53512.

### CODE REVIEW

File: src/test/ui/imports/issue-53269.stderr [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+   |
+LL | macro_rules! mac { () => () }
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+note: `mac` could also refer to the macro imported here
+  --> $DIR/issue-53269.rs:16:9
+   |
+LL |     use nonexistent_module::mac; //~ ERROR unresolved import `nonexistent_module`
+   |         ^^^^^^^^^^^^^^^^^^^^^^^
+
+error[E0601]: `main` function not found in crate `issue_53269`
```
Comment (oli-obk) [inline src/test/ui/imports/issue-53269.stderr]: Please remove the noise by adding a `main` function

File: src/test/ui/imports/shadow_builtin_macros.stderr [hunk 2]
```diff
@@ -1,13 +1,18 @@
...
    |
-LL |         macro_rules! panic { () => {} } //~ ERROR `panic` is already in scope
+LL |         macro_rules! panic { () => {} }
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 LL |     } }
 LL |     m!();
    |     ----- in this macro invocation
-   |
-   = note: macro-expanded `macro_rules!`s may not shadow existing macros (see RFC 1560)
+note: `panic` could also refer to the macro imported here
```
Comment (oli-obk) [inline src/test/ui/imports/shadow_builtin_macros.stderr]: This refers to "here", but there's nothing shown

File: src/librustc_resolve/macros.rs [hunk 3]
```diff
@@ -556,14 +534,15 @@ impl<'a, 'cl> Resolver<'a, 'cl> {
...
     // expansion and import resolution (perhaps they can be merged in the future).
-    pub fn resolve_lexical_macro_path_segment(&mut self,
-                                              mut ident: Ident,
-                                              ns: Namespace,
-                                              record_used: bool,
-                                              force: bool,
-                                              is_attr: bool,
-                                              path_span: Span)
-                                              -> Result<MacroBinding<'a>, Determinacy> {
+    crate fn resolve_lexical_macro_path_segment(&mut self,
```
Comment (oli-obk) [inline src/librustc_resolve/macros.rs]: change to block indent to reduce future churn and return type breakup

**Review by oli-obk — CHANGES_REQUESTED** [review]

**Comment by petrochenkov** [comment]
Updated.

**Review by oli-obk — APPROVED** [review]

**Pull request merged**